### DISCUSSION TITLE
API shorthand definition

### DISCUSSION THREAD

**Discussion by thelinuxlich** [ecyrbe/zodios#143] [Ideas]
I feel like the API for declaring parameters could be simplified, it seems a bit verbose:

```js
apiBuilder({
  path: '/marketplaces/:id/transactions',
  method: 'get',
  alias: 'getMarketplaceTransactions',
  // if it were a POST
  body: {
    .....
  },
  queries: {
      limit: z.number().optional()
      offset: z.number().optional()
      'date_range[gte]': z.string().optional()
      'date_range[lte]': z.string().optional()
  },
  headers: {
    Authorization: z.string()
  },
  // if the user really wants description for openAPI:
  description: {
     queries: {
        limit: 'A parameter to limit the number of rows returned'
     }
  },
  response: transactionsSchema,
})
```
Besides being less verbose, the parameter types have the same name as the objects in the API client (queries, headers) so it's intuitive

**Comment by ecyrbe** [comment]
Sorry, the design of endpoint definition is now stable and won't change as this would be a breaking change for all users.

But you could create a an helper function that transform the definition for you in current format.

**Comment by ecyrbe** [comment]
you can take a look at this if you want to transform it : 
https://github.com/ecyrbe/zodios/blob/main/src/api.ts#L84

**Comment by thelinuxlich** [comment]
By the way, asApi and apiBuilder seems a bit redundant, when should someone choose asApi?

**Comment by ecyrbe** [comment]
It's really a convenient choice.
with `apiBuilder`, since you create one endpoint at a time, you have better autocompletion when creating endpoints.
`asAPI ` on the other end will fail to autocomplete sometimes (typescript fails to understand we are constructing a tuple and instead thinks we are constructing an array).
So i'll recommand using `apiBuilder` when you are creating an api definition, and use `asApi` when combining multiple definitions.

```ts
// in one file
const userAPI = apiBuilder(...).build();
// in another file
const blogAPI = apiBuilder(...).build();
// in another
const fullAPI = asApi([...userAPI, ...blogAPI]);
```

**Comment by ecyrbe** [comment]
moving this issue discussion. if someone wants to contribute to it.

**Comment by thelinuxlich** [comment]
I wrote a helper, it works on runtime, but at compile-time I don't have the aliased methods :(

```ts
import { z } from 'zod'
import { makeApi, makeParameters, ZodiosEndpointParameters } from '@zodios/core'

type Try<A, B, C> = A extends B ? A : C

type NarrowRaw<T> =
  | (T extends Function ? T : never) // eslint-disable-line
  | (T extends string | number | bigint | boolean ? T : never)
  | (T extends [] ? [] : never)
  | {
      [K in keyof T]: K extends 'description' ? T[K] : NarrowNotZod<T[K]>
    }
type NarrowNotZod<T> = Try<T, z.ZodType, NarrowRaw<T>>
type HTTPMethods = 'GET' | 'POST' | 'PUT' | 'DELETE'
type MethodAndPath = `${HTTPMethods} ${string}`
type APIEndpoint = {
  alias: string
  response: z.ZodType<unknown, z.ZodTypeDef, unknown>
  queries?: Record<string, z.ZodType<unknown, z.ZodTypeDef, unknown>>
  headers?: Record<string, z.ZodType<unknown, z.ZodTypeDef, unknown>>
  body?: z.ZodType<unknown, z.ZodTypeDef, unknown>
}
type APIConfig = Record<MethodAndPath, APIEndpoint>

export const api = (config: APIConfig) => {
  const endpoints: ReturnType<typeof makeApi> = []
  for (const [key, value] of Object.entries(config)) {
    const [method, path] = key.split(' ')
    if (method === undefined) throw new Error('Missing method')
    if (path === undefined) throw new Error('Missing path')
    const queries: Array<ReturnType<typeof makeParameters>> = []
    const headers: Array<ReturnType<typeof makeParameters>> = []
    const makeParams = (
      type: NarrowRaw<'Query' | 'Body' | 'Header'>,
      container: Array<ReturnType<typeof makeParameters>>,
      obj: Record<string, z.ZodType<unknown, z.ZodTypeDef, unknown>>
    ) => {
      for (const [_key, _value] of Object.entries(obj)) {
        container.push(
          makeParameters([
            {
              type: type,
              name: _key,
              schema: _value,
            },
          ])
        )
      }
    }
    if (value.queries !== undefined) {
      makeParams('Query', queries, value.queries)
    }
    if (value.headers !== undefined) {
      makeParams('Header', headers, value.headers)
    }
    const endpoint = {
      method: method.toLowerCase() as Lowercase<HTTPMethods>,
      path,
      response: value.response,
      alias: value.alias,
      parameters: [
        ...headers.flat(),
        ...queries.flat(),
      ] as ZodiosEndpointParameters,
    }
    if (value.body !== undefined) {
      endpoint.parameters.push({
        type: 'Body',
        name: 'body',
        schema: value.body,
      })
    }
    endpoints.push(endpoint)
  }
  return makeApi(endpoints)
}
```

Execution:

```ts
import { api } from './_apiHelper'
const standardQueryParams = {
  limit: z.number().optional(),
  offset: z.number().optional(),
  'date_range[gte]': z.string().optional(),
  'date_range[lte]': z.string().optional(),
}
export const zoopAPI = api({
  'GET /v1/marketplaces/:id/transactions': {
    alias: 'getTransactions',
    queries: standardQueryParams,
    response: transactionsSchema,
  },
  'GET /v1/marketplaces/:id/sellers': {
    alias: 'getSellers',
    queries: standardQueryParams,
    response: sellersSchema,
  },
})
```

**Comment by thelinuxlich** [comment]
Must be some TS trick above my league

**Comment by thelinuxlich** [comment]
Not showing the aliases yet, but I think I'm almost there:

```ts
import { z } from 'zod'
import { makeApi, makeParameters, ZodiosEndpointParameters } from '@zodios/core'

type HTTPMethods = 'GET' | 'POST' | 'PUT' | 'DELETE'
type MethodAndAlias = `${HTTPMethods} ${string}`
type APIEndpoint = {
  path: string
  response: z.ZodType<unknown, z.ZodTypeDef, unknown>
  queries?: Record<string, z.ZodType<unknown, z.ZodTypeDef, unknown>>
  headers?: Record<string, z.ZodType<unknown, z.ZodTypeDef, unknown>>
  body?: z.ZodType<unknown, z.ZodTypeDef, unknown>
}
type APIConfig = Record<MethodAndAlias, APIEndpoint>
type ParameterPath<T, K, V> = V extends z.ZodType<
  unknown,
  z.ZodTypeDef,
  unknown
>
  ? {
      type: T
      name: K
      schema: V
    }
  : never
type ValueOf<T> = T[keyof T]
type APIPath<Name, Info> = Name extends `${infer Method} ${infer Alias}`
  ? {
      alias: `${Lowercase<Method>}${Capitalize<Alias>}`
      method: Lowercase<Method>
      path: Info extends APIEndpoint ? Info['path'] : never
      response: Info extends APIEndpoint ? Info['response'] : never
      parameters: Info extends APIEndpoint
        ? [
            ValueOf<{
              [K in keyof Info['queries'] as string]: ParameterPath<
                'Query',
                K,
                Info['queries'][K]
              >
            }>,
            ValueOf<{
              [K in keyof Info['headers'] as string]: ParameterPath<
                'Query',
                K,
                Info['headers'][K]
              >
            }>,
            ParameterPath<'Body', 'body', Info['body']>
          ]
        : never
    }
  : never

type API<C extends APIConfig> = Array<
  { [K in keyof C as string]: APIPath<K, C[K]> }[string]
>

const capitalize = (str: string) => {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export const api = <Config extends APIConfig>(config: Config) => {
  const endpoints: ReturnType<typeof makeApi> = []
  for (const [key, value] of Object.entries(config)) {
    const [method, alias] = key.split(' ')
    if (method === undefined) throw new Error('Missing method')
    if (alias === undefined) throw new Error('Missing alias')
    const queries: Array<ReturnType<typeof makeParameters>> = []
    const headers: Array<ReturnType<typeof makeParameters>> = []
    const makeParams = (
      type: 'Query' | 'Body' | 'Header',
      container: Array<ReturnType<typeof makeParameters>>,
      obj: Record<string, z.ZodType<unknown, z.ZodTypeDef, unknown>>
    ) => {
      for (const [_key, _value] of Object.entries(obj)) {
        container.push(
          makeParameters([
            {
              type: type,
              name: _key,
              schema: _value,
            },
          ])
        )
      }
    }
    if (value.queries !== undefined) {
      makeParams('Query', queries, value.queries)
    }
    if (value.headers !== undefined) {
      makeParams('Header', headers, value.headers)
    }
    const endpoint = {
      method: method.toLowerCase() as Lowercase<HTTPMethods>,
      path: value.path,
      response: value.response,
      alias: method.toLowerCase() + capitalize(alias),
      parameters: [
        ...headers.flat(),
        ...queries.flat(),
      ] as ZodiosEndpointParameters,
    }
    if (value.body !== undefined) {
      endpoint.parameters.push({
        type: 'Body',
        name: 'body',
        schema: value.body,
      })
    }
    endpoints.push(endpoint)
  }
  return makeApi(endpoints) as API<Config>
}
```

**Comment by thelinuxlich** [comment]
I was able to make my type extend ZodiosEnpointDescriptions but it's erroring when passing through Narrow<T>, do you know why that could be happening?

**Reply by ecyrbe** [reply]
Sorry for the delay. Yes, Narrowing can make errors because at the time of evaluation, your types are partially resolved and will only be fully resolved at call site.
So one trick is to just tell typescript to shut up, that you know what you are doing and add a comment `// @ts-expect-error`.
Then it should hopefully work.

**Reply by thelinuxlich** [reply]
Almost there, don't know what I'm still missing 
https://www.typescriptlang.org/play?ts=4.8.2#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFCiSyJlxwBaJwEAzgDS1wgCGA1gKYBBMMA50eAgArco3EPxj8o7To2LMWAUQB2xSMG0xps+YuViuffjr0QDMC9xEAhAK7AANsSVk0nbLgEAAJE6qwA9ADG0PzkZPwAHlTwMACeYPxwAHIyOADuADwAKgB8cAC8cEVQqcVscADaALr1OVD5WRAwasUlJWRpGVU1BYL1zvUAwmWVgnCJiroscM5wAPxwcwBccJNkA+mZbfkAStyFpRWcAD5wABRF8wmLxMsAYq7akTDM2utVcB22n4ADclABKODhcLzFgeewAWnULG4ACMPPwEfDgTd7o8FvwlnAWDAoAYAOZwW7aVwgVFKKlwVHAcn2RmoiAQDHcP4bR5A0EQ3EPJ4vZbNf4SgVgqDg3EITh0OgNADScAMcAEqQgmCqTR2aoJRPw3hYkTJYB+EG0+H+RVVTUB2VyEDynW6JGKDv6SvQ+0GRxdbq6PUulWqtSK9WQADo1EVDq0g2cLn0DkMABJFIqSACyCgAFiRlpV8ABxTRFW23fCSADyAGUq4zawBVZs1gAimgAMpXNNWCJJBEVJhnyAG4AB9LM5-MwIuvCpwHuupSRbgsfgFWd5wvF-qT+eLwS6QTwzfLgAGABIELvj8WMHeSWTtOS0Ff05lBJIAJI2PohjLgqdBgNwC47K+FKcFA-AsJA2hbjssbxocBSfLw2iuto0ZxiQCYZJ2-CYPUmHYXk2g+nAACOrhKMA8FrDsJz8NEUDEAU0HvnhaEZBh2hYThvEEYcxGkXA5E4WmdAFvw3DeMozFwKx7Gcdx5IicQhHblJlFaTp4lkYJFFUdRHLEKkymoaJ-F6bhKD4dpYkkcZQmUf0fiTr+f6TNamAssuqnQJxj7EKe4UXuwmz-oBdiGIehxwMYcgKEo0gLnUcAqvUABqMxwLloqEkuNnOfxnD2RYZWGa5lUmThZDURsoG+nAAY7EUipKtoqUGsVRIaf8ao7Pg5BtcSkRyTwOy5d1fh0NKPjef+GUFgUOTyPUf7aJgEAFZtmRGkut4IAYmAMmFz5nbtDLnsAm6fpwLXdXQ3BRTsp2rnk66btuYUlGgd6TE4wAwO9wDINu92boDX4TaYi47N9v1bgUAMDUuM7ZnuC7Fv8KNQBuaMY6N5IKONbXgZBcA7XtmPLD5cVshsdMQA0+DUwW+COktUCvXAcEIdayG07tEAMzFAG6EB8Cs+LHNC4hW4806wIygL4EmGlyg7GzktMzL8UwALdAbA0pttbcrUTbbjRqhqWo6mLe0c3RDHwTz+rJTIqVmGtBSW3bdD4AAivRNT4BYwe2zlQfB2zbsR4xLA8w68cTdRMd0GgDRO7qif4O7ZKe00TQZ4yNvZ8qDt-PnLvs-gckKUoqdlzsKWmOlEHrRXE34Bm8mKVHfdtXH1cTYXzeKW36cT0qWfZ7n9dT0PrdexXtydzrAf4M4JCpFHBAWYf20K-gJ884vE3l3bfPzZw9-flLBSTAb-5+btLIFQ0SCquqdd+Dal1JMb2PkA45V2N6dAecgHO1AeXMg0QkLwA3CIcG8IobLjuK+KCpIKSQnKGUVqcEYCuCgH8V8MYpoyEEDAO4AAGcEMYYAQFbGADIUAQZbjuJCAA1MSUkMY4TAEiPwO4ABGOUfgkEi3gKDZcr9-KBWOozD+yjyQlDuMggK5IdjHFdEor+mjCHEM4Mgkk8wjb2BLI0W+WBoD3AsfAWBqR6ggnevRR0zs6yogAFZsRgDGQk+D4LaI0eCSEVdnGNERiQeoENNyOkqFqYRYB4T0NtPgOUvpgC6juHE4gFRyiVE+N4AKwJiCQgXPkOAwI8hwE0O0aAdx8C5mACwFgFIuD7lIDkpUeT7iJJLCUySugSIGH4FU9qBZan1Mac0qArT2mdO6cM7J3UYnFxTjsQQ7RuC1FYmQihOkCgBmdhIfg28zAsD6MuZomy5FwGnq3XZ+zDkKHIdoU55zdSXOua3O5lQHm+hif832IBbF3AFh1Ag4clCHxbPvSyg4B5rygCPCayDwaTKgG82QHzjnfPQr8ywUgIU61uSUaOSoID+JYmxEKXF8E8UcnxXSDV9JstsvwIyYz3JmW6qYmgE09pQCcU8hoU4tT1CnB4jwXi4A+P8YE4JhgS4sDuHSvxkSRW22xdwXFMYwCuBYAWaFMdwXaxuXcC21cq4x1hQGGlMderyB2NKoBLrg5mmmtwD18r6LeommgYNSomj9ImpGpUC0Y3dUGXcQN-AYzbPgnAAAhKMspEzKlRIFlauQmqw4R1PrRZO8F3GeOTamlg0bY10ATUmmMLzlAZqzeMipUy80IysJ3Itg8W4YvqC26KTaR11seSgqxtg2SVAdT0vGxAdiFJYRAQm3DxGQkvITYm24Hy9Nud6rmOwm1c29UrEW-AT1VpjBepC-BvXDOXb01d66-q8LgAItBYMIZQzuMM8ER6KU3J2Ha22MYIMjpjJgDwEFeFhogym8tLBoOwfoYBgWjpLxqA0MzQwALzAP1yfkptJ822lI7ZM6Z86Ykn07suAtXdlC2stvO30sK94H0xcHN1V7j5cbDXQX1-AZpwFIwfMNobMPRroCVWWxrgOt2Naa81iG6MQtQ3ByJRGlRyeNihk1Zq7h6fsP02NpCvnTtlssS8PkjG6M8vsGJJJiCMAgGAHyy5QZuE8IpO4rVj0EHCCCCR4QeBQAEDAdJ3AxEsHCFsYAxBwikh5Cib4vxU6OA+gQcmMBqipZi1aJC3HClkwphYLWftXmNG6vOjj8LI7et46NeEIAwbcd9MJ0TsYaR0iULwmM7mivvXg-NGldXDijQa6fAWzWCA6kwFuGAHWlRdf9Y5Xr9IlnMKG78EbGHfRSboC0WC8FlZ8djBpUbaBmGoncF4XhTmnlEHc55yooN-OcHLJWdqsgkKFYy-gHYVdAv4GC6F8LkXouxfi4l5Lf20tFcy91GtwPLatbBihGMm3+s7ctHtjwo27YLaW1jnH23Bv4+tPtmlR2lR3tFpdll5JHs518HKWRU6XtgGXPMnDrALUEALDAKLLAtjQlBjGbnMY8jI-EFYYQwA7jc58jGHgYBlcVDKJcvD9DkCRIwxz5zMBXOch55UPnTABdfeF6L8X4RJfS9l9xlzbmPP-jIEbkWXJk0eAgCz0l0vcv5f++lkWcoLE+5jH7gPhxnau7NzGYPCOAfh7IEAA

**Reply by ecyrbe** [reply]
Hello, i'll take a look once i can reach my computer.
Thanks for consuming so much time on this.
I know doing some advanced typescript can be hard

**Reply by thelinuxlich** [reply]
It's been hell because some things you might expect to work because JS doesn't work in TS-land. But it's cool, I learned about infer and how to use keyof in the process! :)

**Reply by thelinuxlich** [reply]
I finally did it with the help of `ts-toolbelt`:

https://www.typescriptlang.org/play?ts=4.8.2#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFCiSyJwBaJwEAzgDRwgCGA1gKYCCYYG068AChygcQPGDyis6DZgFEAdsUjBVMcZOmz5bDoIBCAV2AAbYnPRYcefAAEixRkwD0AY2g9yFcGh4JAAZNgBVO2xcAhgmAFoYCAhLACMeSxh-HgAPKngYAE8wHjgAOQkcAHcAHgAVAD44AF44OqhC+rYAbQBdNgqoarKIGHpieoaGsiKSto6avjYTNgBhJta+OFzZdSY4EzgAfjgtgC44VbIZ4tLB6oAlDlrGlrI4OAAfOAAKOu2crtiPsAGJmVReGCMVTHNpwC6qHgANzkAEo4B4PNsmJYtDB4m4mBxUpYePFcYj3l9fv8djw9nAmDAoFoAObU1RmEDpKDU1LAVl4vnJUkcGEnf4I5FoqnfP4AoH7Pqw5VSlFQVGyxBUj4fboAaTgWjgvEKEEwbV6F0NdIZ+BsTC8LLAUIgqnwsLqBt68PKlQgVRGYxI9W90116Guszu-sDo3Gkxa806dTYyAAdOM6rcBrGni8plHbnAABJ1OqiACyMgAFiR9q18ABxZR1D3ffCiADyAGU29TO+F+x2ACLKEKt5TtgiiPh1VYl8jRuAAfTLFerMDrwKTIQDci8HCYPBq66rtfr02Xm+3fHUfFxR6TAAMACQIM83+sYd9MlmqVk0GfG45j4UQAEk1A0CAhVaBAqTADgtwuP82SpKAeCYSBVGPC4MyzW4anBLhVADVQ00zEhsxKEceEwNhiNIqpVHDOAAEczDkYBMKOC4Hh4HwoAmVCAIogiSiI1QSLIsSqNuWj6LgRiyMLD4ax4DgbHkXi4H4wThOZNlZOIaiT2U5jjNMhSGKkpiWNY1ISEKHT8LkiTzPIlBKJM+S6Js6TmOmNAQNKMDwNWN1MAFJM9OgCYv2IO9EsfBQwqgzRtCvYtdCkGQ5HELcujgfU2AANQ2OBSoVekd1cnyJKpDyWCpOqrL8xrbLIshWJOeCIw+aMLjqHVdVUXLrWqhkRPZE5DQufByH6xkvHUzgLlKkbgo+NU5BC04IIKmsagqaQ2HA1RMAgCqTtKW0dzfBAtEwWwEp-R6LtsB9gCPICqV6kaPg4FKLgevcqgPI8TwSho0HfVZjGAGAgeAZATy+o8YeApb9G3C4wYh48amhyadzXctzy3etYXxqBD0J4n5tZGRFv6xDkLgc7LpJ-Y0vUDL4BOTmIG6fA2ZrfAfR2qAAbgDCsLdXCOYuiBuf2yC+Zg7RYSFkW5ew48Jd9RF1RlxC9Dy+QLiF1XeegvEZY+E4QnTCKISQmoHf653XcPGAPaWgO4HCdMQmAJku0wf3A4Dvro8Dg0jRhU1zSVy6RY4rjMIlq04By-R8qQo7Pbjj58AARU4jp8Gakvo5K4uS51-AM5ZLPem9BvA9Y2uPjQbpk4tJuW+4phs873UGhrnvg9D8PI-HiNY573UE+NAfU+F-B1M0uRR96HO84tw6o+XiN8BLDStOrheA-r0-+qb7etL3jv74+bue779fH8v3ex+Xj+Xcp5xybo5YghRDZ3X2K1QiTUvLiR4NZJSnVAo3xON0Q+Bhj74BME5auBAwEQLOsrEWhCJYNF6DfC4fQG6AN1FLTaVIGF7TCjUVYNsIKuyiqyCq3s3S+yjjPMOMAI41CQKvJOPAzQWlWDne4AZFgHULjUEqlwwxNC-lIlOsjJ5UhodMMgPgcLwEPIIJGuJUZJh+H+FChkALomaE0WOGEYBmCgDCP86YVoSD4DAH4AAGVE6YkjhDACUKA8Njw-HRAAakZMydMOJgBeB4D8AAjJqYKhiFbwARkmNhkVopQLVlwgUDQfhGO4RceRtRSk8IcU4qkRimTbA1niBscAaEfEuryCpOTOmmjYEiIGnEfQpy7KkAAVgJGA6Z6SGUwn0i6ApUToiXnAZp8Bug4xIEYFKPpWimkSWAXEfiPT4E1BGYAFofg7OIC0ZorRwQ2CioiYg6ItzVDgIiKocBlBDGgD8fAlYw5MDZOwC8pBLm6mub8ZGT5HlPPUHRLQPB3lwE+QGb5PBfn-JwFAIFIKmBgoAnAeFo9oUfE2exSuI8Lh8CGBwTo-FXHuNMjUaMKcRA8EwbvKYSYum6mpU-Xe9LGXMpkG41Q7LOUWm5by+Q-LWiCqpf0+VEgpAdJ+DLQaBAK5yAgQOXB4Dpzn1-lAa+S0jFI1RVAMVkgJWsulYRWV7BuA8o1fnRVuiloQCmXxAScUajTUsrAlBnkYE0T8sggK9kRoNO1EtHpvxqXdBXIM1cwzLCjLgOMqZMy5naFbkwH4frJmrMTQHa1HBbXpjAGYJgNZtVx3VebAwJbujj3WXHXV0ZgElzGtIC46apH9rjo6VaHBh1Zs4mOgOaA50Rl6JS-qK7e6MKuTcmdPB0zD0wnAAAhI85BLzUXou7W6sQnqS3l0rkQmlmcFDbt3bSzCK6towq3SMndIr5CHuPc8lFby1ky1bZqoFF8d4WrYL+p9370ywffSNalNV+ZJgvXci4dzgkQBppE1J6Inw0zpieT8kKmA+tZoXC4z6xZzr1grHgNH4MMZwjwOd5KsOQpw3hyG0S4BxNMYjZGqMfjktRHOs2uV23UM9umeTsH0yYEsEhaJi64DyZfY+pTKm-ESZlj6J84x3DpU1joT1FtWAbs-b8Z9hD-1ItPcBytEZqWELzkmMDXqO2ewvRGXVOC8HqY+IOpjBCgsNwnTwNacA7NOXUwugza7Wl220HWiz7a60NqbZp9znqdOqdWdZj4qGzNMCy42n4pW8TQo-S4qVKX+b7CfKwupQVrjUqZMQegEAwBhSTAjcwVgtI-FjmLeaHgkRpI8JwKAvAYAnI4CkzwZxgDEA8MyMURJITQlHlPTjBAmYwHaFtpbrocKWohZTYgjNmZTyk952Ti8dW3Hmvqquc7QvzVxCARGl2IxRZixmTk3I5DRPTL187QM1ObX7RegL72iEyy+wQc0mBjxZDnYDqdXkQc8nB5D6E0P9MRkSx8fo6FML6zCxmaaMO0BBNSBYaw0SOv9KIL1-rrQEajapM2VsGLJA4TO7t-AFx1njYIJN6bs35uLeWx4Vb63NvC52wrf7D7i3i89j9xGeF0x47B0EwnbpifqbRxj-XhuCXG5dETywMPSf9tY4rWndjWSs-XQztnxiUDJDAEmH5ig3DMGbQQGsMAFtMDOJiBG6YOdgHTFUPbVJuUCGAD8BPYV9Oamyb7rrPWA+tCD8Z0PfOI9R5jx4OPCek8p4+AX-3YUyC5+aSkHdlgIAe9dbXo7J3VfnaYJqNvpJ0yd+77cFOjfevpj70L7bg-NRAA

**Reply by thelinuxlich** [reply]
Only thing missing now is the params object 
![image](https://user-images.githubusercontent.com/5043/192370543-fe330d83-f64f-4061-b70b-7fb4fbc0f0da.png)

**Reply by ecyrbe** [reply]
This is rocket science level. So glad you are succeeding.

**Reply by thelinuxlich** [reply]
If I declare the path as const it works, I don't know how can I get rid of this for now 
![image](https://user-images.githubusercontent.com/5043/192372778-30bc373b-2aae-4821-b47d-4f83580da473.png)

**Reply by ecyrbe** [reply]
is this for puting a predefined id to evoid passing this path parameter everytime ?

**Reply by thelinuxlich** [reply]
yes

**Reply by ecyrbe** [reply]
in ts playground, with your last impl, it seems, to work without `as const`. but maybe i'm looking at it in the wrong way.

**Reply by thelinuxlich** [reply]
@ecyrbe solved it and published the lib here: https://github.com/thelinuxlich/zodios-api-shorthand

**Reply by ecyrbe** [reply]
@thelinuxlich well done! i will add it to the zodios ecosystem !

**Reply by ecyrbe** [reply]
done: https://www.zodios.org/docs/ecosystem

**Reply by thelinuxlich** [reply]
Thanks, glad I could solve this and learn a lot in the process
